File: r/compound_api_interactive_demo.R
# compound_api_interactive_demo.R — Interactive HTML for compound C API paths
#
# Same UI as python/compound_api_interactive_demo.py (shell:
# python/compound_api_interactive_shell.html). Frames are computed in R via
# ic_interval_1d, r_scs_winset_2d_export_boundary, uncovered_set_boundary_2d,
# and r_scs_voronoi_cells_2d (heap-backed in C).
#
# Shared build/validation: r/tests/testthat/compound_api_interactive_common.R
#
# From repository root:
#   SCS_LIB_PATH=$(pwd)/build DYLD_LIBRARY_PATH=$(pwd)/build Rscript r/compound_api_interactive_demo.R
#
# Optional: COMPOUND_INTERACTIVE_STRICT=1 exits non-zero if data checks fail.
#
# Writes tmp/compound_api_interactive_r.html and opens it in the browser
# (set OPEN_BROWSER <- FALSE to skip).

OPEN_BROWSER <- TRUE
OUT_NAME <- "compound_api_interactive_r.html"

root <- getwd()
r_pkg <- file.path(root, "r", "DESCRIPTION")
here_pkg <- file.path(root, "DESCRIPTION")
if (file.exists(r_pkg)) {
  repo_root <- root
  pkg_root <- file.path(root, "r")
  devtools::load_all(pkg_root, quiet = TRUE)
} else if (file.exists(here_pkg)) {
  pkg_root <- root
  repo_root <- dirname(root)
  devtools::load_all(pkg_root, quiet = TRUE)
} else {
  stop("Run from repo root (r/DESCRIPTION) or from r/.")
}

common <- file.path(pkg_root, "tests", "testthat", "compound_api_interactive_common.R")
if (!file.exists(common)) {
  stop("Missing common helpers: ", common)
}
source(common, local = FALSE)

shell <- file.path(repo_root, "python", "compound_api_interactive_shell.html")
if (!file.exists(shell)) {
  stop("Missing shell template: ", shell)
}

payload <- build_interactive_payload()
vc <- list(checks = payload$meta$build_checks, all_ok = payload$meta$build_all_ok)
report_lines <- format_interactive_validation_report(vc)
message(paste(report_lines, collapse = "\n"))
strict <- identical(Sys.getenv("COMPOUND_INTERACTIVE_STRICT", ""), "1")
if (isTRUE(strict) && !isTRUE(vc$all_ok)) {
  quit(status = 1L)
}

json <- jsonlite::toJSON(
  payload,
  auto_unbox = TRUE,
  null = "null",
  digits = 10
)
tpl <- paste(readLines(shell, encoding = "UTF-8", warn = FALSE), collapse = "\n")
parts <- strsplit(tpl, "__DATA__", fixed = TRUE)[[1L]]
if (length(parts) != 2L) {
  stop("Shell template must contain exactly one __DATA__ placeholder.")
}
html <- paste0(parts[[1L]], json, parts[[2L]])
html <- gsub("__SOURCE_TAG__", "R bindings", html, fixed = TRUE)

out_dir <- file.path(repo_root, "tmp")
if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)
out_path <- file.path(out_dir, OUT_NAME)
cat(html, file = out_path, sep = "")
message("Wrote ", normalizePath(out_path, winslash = "/"))

if (isTRUE(OPEN_BROWSER)) {
  utils::browseURL(out_path)
}

File: python/compound_api_interactive_demo.py
"""compound_api_interactive_demo.py — Interactive HTML demos for compound C API paths
=====================================================================================

Builds a self-contained HTML file (canvas + tabbed panels + sliders). Each slider
step is precomputed in Python via ``ic_interval_1d``, ``scs_winset_2d_export_boundary``,
``uncovered_set_boundary_2d`` (heap), and ``scs_voronoi_cells_2d_heap``.

The page template lives in ``compound_api_interactive_shell.html`` (shared with the
R script ``r/compound_api_interactive_demo.R``). Shared build/validation logic is in
``compound_api_interactive_common.py``.

Run from the repository root::

    export SCS_LIB_PATH=$(pwd)/build
    python python/compound_api_interactive_demo.py

Use ``--strict`` to exit with code 1 if embedded data checks fail (for CI).

Writes ``tmp/compound_api_interactive_py.html`` and opens it in the default browser
unless ``OPEN_BROWSER = False``.
"""

from __future__ import annotations

import sys
import webbrowser
from pathlib import Path

from compound_api_interactive_common import build_payload, format_validation_report

OPEN_BROWSER = True
OUT_NAME = "compound_api_interactive_py.html"


def main() -> None:
    strict = "--strict" in sys.argv
    py_dir = Path(__file__).resolve().parent
    root = py_dir.parent
    shell = py_dir / "compound_api_interactive_shell.html"
    if not shell.is_file():
        raise FileNotFoundError(f"Missing template: {shell}")
    out = root / "tmp" / OUT_NAME
    out.parent.mkdir(parents=True, exist_ok=True)
    payload = build_payload()
    meta = payload["meta"]
    report = format_validation_report(meta["build_checks"], meta["build_all_ok"])
    all_ok = meta["build_all_ok"]
    print(report)
    if strict and not all_ok:
        sys.exit(1)

    import json

    html = shell.read_text(encoding="utf-8")
    html = html.replace("__DATA__", json.dumps(payload)).replace(
        "__SOURCE_TAG__", "Python bindings"
    )
    out.write_text(html, encoding="utf-8")
    print(f"Wrote {out}")
    if OPEN_BROWSER:
        webbrowser.open(out.as_uri())


if __name__ == "__main__":
    main()
